runtime/caml/domain_state.h

```diff
@@ -0,0 +1,56 @@
+/**************************************************************************/
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*      KC Sivaramakrishnan, Indian Institute of Technology, Madras       */
+/*                Stephen Dolan, University of Cambridge                  */
+/*                                                                        */
+/*   Copyright 2019 Indian Institute of Technology, Madras                */
+/*   Copyright 2019 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#ifndef CAML_STATE_H
+#define CAML_STATE_H
+
+#include <stddef.h>
+#include "misc.h"
+#include "mlvalues.h"
+
+#undef young_ptr
+#undef young_limit
+#undef young_start
+#undef young_end
+#undef local_roots
+
+/* This structure sits in the TLS area and is also accessed efficiently
+ * via native code, which is why the indices are important */
+typedef struct {
+#define DOMAIN_STATE(type, name) CAMLalign(8) type name;
+#include "domain_state.tbl"
+#undef DOMAIN_STATE
+} caml_domain_state;
+
+enum {
+  Domain_state_num_fields =
+#define DOMAIN_STATE(type, name) + 1
+#include "domain_state.tbl"
+#undef DOMAIN_STATE
+};
+
+/* Check that the structure was laid out without padding,
+   since the runtime assumes this in computing offsets */
+CAML_STATIC_ASSERT(
+  sizeof(caml_domain_state) ==
+   (Domain_state_num_fields
+   ) * 8);
+
+CAMLextern caml_domain_state* Caml_state;
+
+#include "compatibility.h"
```

Comment:
Note that this doesn't work if `compatibility.h` is actively being used (i.e. it's being included in user code). I'm not sure what the impact is (as in, what code out there relies on `compatibility.h` in order to access the five things which have changed), but this comment is just to track it, and note that it's on my list to propose a fix.

(I don't think it has to block merging)

Reply:
I wonder if the easiest thing to do here is to remove the `#include "compatibility.h"` and instead just redefine the undefed macros. `domain_state.h` includes `mlvalues.h` which conditionally includes `compatibility.h` if `CAML_NAME_SPACE` is not defined. Hence, the change will be to delete `#include "compatibility.h"` and replace that with

```c
#ifndef CAML_NAME_SPACE
#define young_start (Caml_state->young_start)                                       
#define young_end (Caml_state->young_end)                                           
#define young_ptr (Caml_state->young_ptr)                                           
#define young_limit (Caml_state->young_limit) 
#endif 
```
We should add comments in `compatibility.h` and `domain_state.h` referring to the other.

Reply:
I pushed a fix for this in https://github.com/ocaml/ocaml/pull/8713/commits/239a667d8b6ca559d7b5fa86a88806adc3397c9f. This should leave the behavior of programs that depend on `compatibility.h` unchanged.